design/20210203.certificate-request-identity.md

```diff
@@ -0,0 +1,165 @@
+these resources must be updated to have parity with the upstream Kubernetes
+certificates API.
+This means that the `CertificateRequest` API type be updated to include the
+following fields in spec, for all API versions:
+
+```go
+	// Username contains the name of the user that created the CertificateRequest.
+	// Populated by the cert-manager webhook on creation and immutable.
+	// +optional
+	Username string `json:"username,omitempty"`
+	// UID contains the uid of the user that created the CertificateRequest.
+	// Populated by the cert-manager webhook on creation and immutable.
+	// +optional
+	UID string `json:"uid,omitempty"`
+	// Groups contains group membership of the user that created the CertificateRequest.
+	// Populated by the cert-manager webhook on creation and immutable.
+	// +listType=atomic
+	// +optional
+	Groups []string `json:"groups,omitempty"`
+	// Extra contains extra attributes of the user that created the CertificateRequest.
+	// Populated by the cert-manager webhook on creation and immutable.
+	// +optional
+	Extra map[string][]string `json:"extra,omitempty"`
+```
+
+### Upgrading
+
+Any `CertificateRequest`s that are present in the cluster at the time of
+upgrade, won't and will never have their identity fields populated. This should
+be acceptable.
+
+All subsequent `CertificateRequest`s created after this upgrade will have their
+identities populated.
+
+### Test Plan
+
+Unit tests will be created is ensure that the [properties](#proposal) are
+enforced.
+
+End to end tests will ensure that different users creating `CertificateRequest`
+resources have their respective identities populated in the fields.
+
+### Risks and Mitigations
+
+There are large security implications if the identity fields that other
+components rely on are wrong, or could be made fraudulent. Special care when
+testing needs to be given to ensure the properties described above are correct
+and enforced.
+
+### Certificate Identity Pass Down
+
+`CertificateRequest`s created and managed via `Certificate` resources will have
+the identity of the cert-manager controller. Though this design doesn't provide
+a solution to components that need to traverse the identity to the original
+user who created the `Certificate` resource, some possible further design
+options are;
+- cert-manager would override the identity fields with the of the identity that
+    created the `Certificate` resource, if cert-manager is the creator.
+- Consumers of the identity will need to be aware that the `CertificateRequest`
+    may be managed by a `Certificate` resource if it is created by cert-manager,
+    and should look at the identity of the `Certificate`.
+- cert-manager manages some signing key pair and populates the annotation of
+    the `CertificateRequest` with a JWT containing the original identity.
+    Consumers can then validate that JWT using a distributed public key, and
+    make use of the identity it contains.
```

Comment:
I personally lean towards the latter option as it does not special-case the cert-manager controller. Thanks for noting down these options though, this will form a great basis for the upcoming conversation on how this _is_ handled.